Implement Adjusted mutual information

christiangnrd opened the pull request:
I may have gone overboard with the code reuse. I can refactor if needed.

christiangnrd committed:
Implement Adjusted Mutual Information

alyst commented:
Thank you for the contribution! I think it would be useful addition.
I have not reviewed your PR in the detail, but there are some things that I have noticed:
 * You introduce the new `adjustedmutualinfo` method, I think it would be better to keep a single `mutualinfo()`, just add the `method` kwarg, and allow method-specifc kwargs (`averaging`, `normed`).
 * I also think it is better to stay conservative with Clustering deps.
   You can move *SpecialFunctions* to the weakdeps (because it brings ChainRules, OpenLibm, LogExpFunctions with it etc) and move the `_mutualinfo_adjusted()` into the `ClusteringSpecialFuncsExt`.

christiangnrd commented:
Thanks for the quick feedback! 

Regarding your first point, isn’t that breaking? And how should I deal with the 4 variants of the adjusted mutual information? Have one symbol for each? (And another that’s just `:adjusted` as default)

And good point about `SpecialFunctions`, I somehow assumed it was lightweight and did not verify.

alyst commented:
> Regarding your first point, isn’t that breaking? And how should I deal with the 4 variants of the adjusted mutual information? Have one symbol for each? (And another that’s just :adjusted as default)

I am thinking about something like

```julia
mutualinfo(a, b; method = :classic, kwargs...) = _mutualinfo(Val{method}, a, b; kwargs...)

function _mutualinfo(::Val{:classic}, a, b; normed::Bool = true)
    ....
end

function _mutualinfo(method::Val, a, b; kwargs...)
    if method[] == :adjusted
        error("mutualinfo(..., method=:adjusted) requires SpecialFunctions package")
    else
        error("mutualinfo(): method=$(method) is not supported")
    end
end
```

and then in the extension
```julia
function _mutualinfo(::Val{:adjusted}, a, b; aggregate::Symbol = :mean)
    ....
end
```

So since the default method value is `:classic`, the old code should not be broken (I don't know if `:classic` is the best name, probably you know better).
I think this is a pretty standard pattern for Julia extensions.
And it allows each method to have its own keyword arguments.
So it will automatically pass `aggregate` kwarg to your adjusted mutual information implementation
(I suggest to rename `average_method` to `aggregate` (taking values `:mean`, `:max`, `:min`, [`:geomean`](https://juliastats.org/StatsBase.jl/v0.19/means.html#StatsBase.geomean)), since it's a more generic term, which I think better fits here)

christiangnrd committed:
Project.toml: Clean up

christiangnrd commented:
@alyst I made the changes you suggested with a few adjustments. The `method` parameter can now take `:classic`, ':normalized`, and `:adjusted`. I did it in a way to preserve backwards compatibility, but I think this is better since the NMI is no less a different variant of MI than AMI is.

As for the value to use for the regular Mutual Information, `:classic` could work. I think I slightly prefer `:original`, but I have no justification other than I like it more, so I'll go with whatever you think is best.

The 1.0 test is failing because `LazyString`s were only added in 1.8. Can I raise the minimum Julia to 1.10 (the latest LTS)? I can't imagine anyone is using an older version of Julia while also updating their environment.

alyst commented:
>  I made the changes you suggested with a few adjustments.

Thank you!

> The method parameter can now take :classic, ':normalized, and :adjusted`. 

I think it is better to raise *ArgumentError* if *method == :normalized* and *normed == false*.
Otherwise you silently give *normed* param priority over *method*, but in this case both *method* and *normed* are non-default, so they must have been explicitly given by the user.

Also, if *normed* kwarg is in the generic `mutualinfo()` function, it has to be supported by all methods, but it looks like for *:adjusted* it will be no-op at the moment.
If *adjusted* would not be supporting *normed*, I think it is better to keep it specific for "classic" method only, and the *adjusted* one will automatically throw *MethodError* if `normed = ` is supplied.

Regarding `:normalized`: you can have this logic in the generic mutualinfo that if *method == :normalized*, you detect it and call `_mutualinfo(Val(:classic), ...; normed = true, kwargs...)`.

> Can I raise the minimum Julia to 1.10 (the latest LTS)?

Extensions will require us to bump Julia requirement to 1.8 (and we can use backward compatibility packages to support earlier versions, probably 1.6 LTS) -- which is reasonable, given how outdated 1.0 is by now.
LazyStrings are nice, but I don't see a real use-case for Clustering.jl where the advantages of LazyStrings will be noticeable.
The negative effect of setting Julia to 1.10+ is that 1.10 is quite recent, and it will exclude many users of older Julia versions from the latest *Clustering.jl*.

christiangnrd commented:
> I think it is better to raise _ArgumentError_ if _method == :normalized_ and _normed == false_.
> Otherwise you silently give _normed_ param priority over _method_, but in this case both _method_ and _normed_ are non-default, so they must have been explicitly given by the user.

`method=:normalized` is the default since previously, `normed=true` was default. I changed things around such that `method` and `normed` can't be used together, and I officially deprecated `normed`. Now, an `ArgumentError` is thrown for any invalid combination of arguments.

> Extensions will require us to bump Julia requirement to 1.8 (and we can use backward compatibility packages to support earlier versions, probably 1.6 LTS) -- which is reasonable, given how outdated 1.0 is by now.
> LazyStrings are nice, but I don't see a real use-case for Clustering.jl where the advantages of LazyStrings will be noticeable.
> The negative effect of setting Julia to 1.10+ is that 1.10 is quite recent, and it will exclude many users of older Julia versions from the latest _Clustering.jl_.

Since I removed the use of `LazyString`, this package works for every version since 1.0. For Julia versions before 1.9, SpecialFunctions will simply be treated as a regular dependency.

christiangnrd committed:
tests: Fix tests for older Julia

Also a couple of clarity fixes

christiangnrd commented on test/mutualinfo.jl:
```diff
@@ -6,13 +6,31 @@ using Clustering
+    @test mutualinfo(a1, a2; method=:adjusted, aggregate=:min) ≈ 0.4348 atol=1.0e-4
+
+    # test errors
+    @test_throws "ArgumentError: `normed` kwarg is not compatible with `method` kwarg" mutualinfo(a1, a2; method=:adjusfted, normed=false)
```
This does not change anything as the error of `normed` should take precedence but I fixed it anyway to make it more clear what's being tested.

christiangnrd commented on src/mutualinfo.jl:
```diff
@@ -14,28 +14,73 @@ function _mutualinfo(A::AbstractMatrix{<:Integer}, normed::Bool)
+    end
+    # Little hack to ensure the correct error is thrown
+    if method === :adjusted && length(kwargs) >= 1 && :aggregate ∉ keys(kwargs)
+        method = :classic
```
This is wrong but I clarified the comment a bit

christiangnrd committed:
CI: Fix codecov and macOS arch

christiangnrd commented:
@alyst This is ready to merge and tests are passing!

Changes I've made since the review:
- I've made some tweaks to comments to make them more clear based on Copilot's wrong comments
- I've fixed tests and code for older Julia versions.
- I've reorganized the PR into 4 commits: Code before today, tweaks from today, cleanup of Project.toml, and 2 fixes to CI to silence warnings.
